lenskit-core/src/main/java/org/lenskit/data/entities/Entity.java

```diff
@@ -0,0 +1,154 @@
+    Collection<AttributeValue<?>> getAttributeValues();
+
+    /**
+     * View this entity as a map.
+     * @return A map reflecting the entity's data.
+     */
+    Map<String,Object> asMap();
+
+    /**
+     * Check if the entity has a field with a particular name.
+     *
+     * @param name The field name to look for.
+     * @return `true` if the entity has a field named `name`.
+     */
+    boolean hasAttribute(String name);
+
+    /**
+     * Check if the entity has a attribute.
+     *
+     * @param attribute The attribute to look for.
+     * @return `true` if the entity contains `attribute` (same name **and type**).
+     */
+    boolean hasAttribute(Attribute<?> attribute);
+
+    /**
+     * Get the value of an attribute.
+     * @param attribute The attribute.
+     * @param <T> The attribute's type.
+     * @return The attribute's value.
+     * @throws NoSuchAttributeException if the specified attribute is not present.
+     * @throws IllegalArgumentException if a attribute with the same name as `attribute` is present, but it is of an
+     *         incompatible type.
+     */
+    @Nonnull
+    <T> T get(Attribute<T> attribute);
+
+    /**
+     * Get the value of an attribute by name.
+     * @param attr The attribute name.
+     * @throws NoSuchAttributeException if the specified attribute is not present.
+     * @return The field value, or `null` if the attribute is missing.
+     */
+    @Nonnull
+    Object get(String attr);
+
+    /**
+     * Get the value of a possibly-missing attribute.
+     * @param attribute The attribute.
+     * @param <T> The attribute's type.
+     * @return The attribute's value, or `null` if it is not present.
+     * @throws IllegalArgumentException if a attribute with the same name as `attribute` is present, but it is of an
+     *         incompatible type.
+     */
+    @Nullable
+    <T> T maybeGet(Attribute<T> attribute);
+
+    /**
+     * Get the value of a possibly-missing attribute by name.
+     * @param attr The attribute name.
+     * @return The attribute's value, or `null` if it is not present.
+     * @throws IllegalArgumentException if an attribute with the same name as `attr` is present, but it is of an
+     *         incompatible type.
+     */
+    @Nullable
+    Object maybeGet(String attr);
+
+    /**
+     * Get the value of a attribute that contains a long.
+     * @param attribute The attribute.
+     * @return The attribute's value.
+     * @throws NoSuchAttributeException if the specified attribute is not present.
+     * @throws IllegalArgumentException if the attribute is present but its type is not `Long`.
+     */
+    long getLong(Attribute<Long> attribute);
+
+    /**
+     * Get the value of a attr that contains a double.
+     * @param attr The attr.
+     * @return The attr's value.
+     * @throws NoSuchAttributeException if the specified attr is not present.
+     * @throws IllegalArgumentException if the attr is present but is not of type `Double`.
```

Comment:
not relevant as written. attribute's identify on type so non-Double would mean that its not present